### PR TITLE
Feature checkls includes settings include where you use

### PR DESCRIPTION
**Pull request by killerbot242**

### CODE REVIEW

**Review by jtv — CHANGES_REQUESTED** [review]
Thanks for continuing the cleanup.  I found one place where I think things could be easier.

File: tools/generate_check_config.py [hunk 1]
```diff
@@ -95,8 +95,12 @@ def generate_cmake(source: Path, snippets: list[str]) -> None:
     dest = source / 'cmake' / CMAKE_CONFIG
     with dest.open('w', encoding='ascii') as stream:
         stream.write(compose_header())
+        stream.write("include(CheckCXXSourceCompiles)\n")
```
[the remark this answers is not in the document]
Comment (killerbot242) [reply tools/generate_check_config.py:98]: agreed, that's another way, and probably better, I will try this out.

File: tools/generate_check_config.py [hunk 2]
```diff
@@ -95,8 +95,12 @@ def generate_cmake(source: Path, snippets: list[str]) -> None:
...
+        stream.write("set(CMAKE_REQUIRED_LIBRARIES pq)\n")
+        stream.write("set(CMAKE_REQUIRED_QUIET ON)\n")
         for snippet in snippets:
             stream.write(CMAKE_TEMPLATE % {'macro': Path(snippet).stem})
+        stream.write("set(CMAKE_REQUIRED_QUIET OFF)\n")
```
[the remark this answers is not in the document]
Comment (killerbot242) [reply tools/generate_check_config.py:103]: agreed, that's another way, and probably better, I will try this out.

[the remark this answers is not in the document]
Comment (killerbot242) [reply tools/generate_check_config.py:103]: FOOT is also used for the "generate_autoconf", where this is not needed, right ? So FOOT is not ok  ...

File: tools/generate_check_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (killerbot242) [reply tools/generate_check_config.py:98]: HEAD is also used for the "generate_autoconf", where this is not needed, right ? So HEAD is not ok ...

**Review by jtv — APPROVED** [review]
Arrrgh.  I had my comment sitting around for days in "Pending" state because it wasn't obvious to me anywhere that it wasn't being shown.  My apologies.  I just found the button that should make it appear.

File: tools/generate_check_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (killerbot242) [reply tools/generate_check_config.py:98]: yes, yesterday your comment became visible to me

**Comment by killerbot242** [comment]
so how to proceed, we add extra complexity, that there is a header for cmake path and the autoconf path, and add a footer for the cmake path and autoconf path ?
Personally I have the feeling this would make the thing more complex.

I am also looking into, removing those 2 or 3 lines all together (except the include, so the discussion on the footer would remain), because I think that:
* CMAKE_REQUIRED_QUIET to off and on has very little added value (still need to test to show the difference)
* CMAKE_REQUIRED_LIBRARIES might not be needed (I still have to investigate)

**Comment by killerbot242** [comment]
on the topic of "CMAKE_REQUIRED_QUIET",  the "all status messages associated with the check will be suppressed." is not used for the check_function_exists
==>
-- Looking for poll
-- Looking for poll - found

It is however used for try_compile, so this is not consistent. Note in many projects you typically see those messages, so one can see what happens, so my advice would be "do not try to hide them".

However, currently (I will create a PR next week or so) we are using try_compile (it is advised not to use that but use the wrappers around it, like "check_cxx_source_compiles".
And that one again can be customized with things like "CMAKE_REQUIRED_QUIET', see : https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html.

When looking at the documentation of : https://cmake.org/cmake/help/latest/command/try_compile.html
you notice nothing is mentioned on CMAKE_REQUIRED_QUIET.
So I tend to conclude that the CMAKE_REQUIRED_QUIET OFF/ON we have is actually useless.

So if ok, I would create a new PR, to remove it, before continuing with this PR (or we can even abandon this PR).

Similar I think that CMAKE_REQUIRED_LIBRARIES also has no influence when using try_compile directly.

Next to that, I just checked all the test code snippets, NONE of them need to link with pq library, meaning the 
set(CMAKE_REQUIRED_LIBRARIES pq) is not needed.


So concluding as a proposal:
1. create PR to remove CMAKE_REQUIRED_LIBRARIES  (since not needed)
2. create PR to remove CMAKE_REQUIRED_QUIET OFF/ON< since no effect due to try_compile, and not consistent, and would suggest even if it would have work not to hide those messages
3. create a new PR, for the include ==> to FOOTER or not TO FOOTER ?
4. create PR to stop using naked try_compile and switch to the advised check_cxx_source_compiles

What do you think ?

**Comment by jtv** [comment]
> so how to proceed, we add extra complexity, that there is a header for cmake path and the autoconf path, and add a footer for the cmake path and autoconf path ? Personally I have the feeling this would make the thing more complex.

Never mind then — it was just a thought.  If it complicates things...  Well, the future tense of "complexity" is "headache."  :-)


> I am also looking into, removing those 2 or 3 lines all together (except the include, so the discussion on the footer would remain), because I think that:
> 
>     * CMAKE_REQUIRED_QUIET to off and on has very little added value (still need to test to show the difference)
> 
>     * CMAKE_REQUIRED_LIBRARIES might not be needed (I still have to investigate)

I'm not the best judge, since I'm not that good at CMake.  (Almost all of this stuff was contributed by others.)

However the `CMAKE_REQUIRED_LIBRARIES` sounds like it should be important..?

**Comment by jtv** [comment]
> on the topic of "CMAKE_REQUIRED_QUIET", the "all status messages associated with the check will be suppressed." is not used for the check_function_exists ==> -- Looking for poll -- Looking for poll - found

> So if ok, I would create a new PR, to remove it, before continuing with this PR (or we can even abandon this PR).

Actually yes I _would_ like to see that output.  I get it from the `configure` script, and sometimes it's helpful — not just for its convenience but also because sometimes somebody happens to notice an unexpected outcome that would otherwise slip under the radar.


> Next to that, I just checked all the test code snippets, NONE of them need to link with pq library, meaning the set(CMAKE_REQUIRED_LIBRARIES pq) is not needed.

Oh — I thought it would affect the build of libpqxx itself as well!

In older versions of libpqxx there were configuration checks for specific libpq features, and so it was sometimes important to link with libpq.  But then one day I found that none of the functions I cared about was missing in any supported version of libpq.  And nowadays there are even feature macros that will let us detect the presence of features at compile time.


> So concluding as a proposal:
> 
>     1. create PR to remove CMAKE_REQUIRED_LIBRARIES  (since not needed)

If it's only needed for the feature checks, then yes, let's get rid of it.


>     2. create PR to remove CMAKE_REQUIRED_QUIET OFF/ON, since no effect due to try_compile, and not consistent, and would suggest even if it would have worked not to hide those messages

Agreed.


>     3. create a new PR, for the include ==> to FOOTER or NOT to FOOTER ?

Never mind that.  It was just a thought.


>     4. create PR to stop using naked try_compile and switch to the advised check_cxx_source_compiles

If it's recommended to avoid `try_compile`, then sure.  It'll be nice to get closer to modern recommended practice.

Thanks again for your work on this.

**Comment by jtv** [comment]
@killerbot242 do we still need this PR?  Or are you working on an updated plan?

**Comment by killerbot242** [comment]
i am working on the updated plan, what is left to do is the move of the include to the checks file (aka adjustment of the generator script), we can try to do this with this PR, or you can reject this pone, and I create a fresh new one.
The latter has my preference.
And we were not gonna done the header thing, right ?

**Comment by jtv** [comment]
Right @killerbot242 — I'll reject this one then.  And as you say, never mind the header/footer idea.

**Pull request closed without merging**